--- a/lib/cpp/src/thrift/protocol/TCompactProtocol.tcc
+++ b/lib/cpp/src/thrift/protocol/TCompactProtocol.tcc
@@ -38,11 +38,13 @@
 # error "TCompactProtocol currently only works if a signed right shift is arithmetic"
 #endif
 
+#ifndef UNLIKELY
 #ifdef __GNUC__
 #define UNLIKELY(val) (__builtin_expect((val), 0))
 #else
 #define UNLIKELY(val) (val)
 #endif
+#endif
 
 namespace apache { namespace thrift { namespace protocol {
 
