--- a/cpp/cmake_modules/ThirdpartyToolchain.cmake
+++ b/cpp/cmake_modules/ThirdpartyToolchain.cmake
@@ -2434,12 +2434,34 @@ macro(build_zlib)
   set(ZLIB_STATIC_LIB "${ZLIB_PREFIX}/lib/${ZLIB_STATIC_LIB_NAME}")
   set(ZLIB_CMAKE_ARGS ${EP_COMMON_CMAKE_ARGS} "-DCMAKE_INSTALL_PREFIX=${ZLIB_PREFIX}")
 
-  externalproject_add(zlib_ep
-                      ${EP_COMMON_OPTIONS}
-                      URL ${ZLIB_SOURCE_URL}
-                      URL_HASH "SHA256=${ARROW_ZLIB_BUILD_SHA256_CHECKSUM}"
-                      BUILD_BYPRODUCTS "${ZLIB_STATIC_LIB}"
-                      CMAKE_ARGS ${ZLIB_CMAKE_ARGS})
+  file(WRITE ${CMAKE_BINARY_DIR}/fix_fdopen_patch.sh "
+  #!/bin/bash
+  set -e
+  echo 'Fixing fdopen macro in zutil.h'
+  if grep -q '# *define[[:space:]]\\+fdopen' zutil.h; then
+    sed -i.bak \$'s|# *define[[:space:]]\\+fdopen|#if !defined(fdopen)\\\\n#define fdopen|' zutil.h
+    echo '#endif' >> zutil.h
+  else
+    echo 'No fdopen macro found, skipping patch.'
+  fi
+  ")
+  file(CHMOD
+    ${CMAKE_BINARY_DIR}/fix_fdopen_patch.sh
+    PERMISSIONS
+      OWNER_READ OWNER_WRITE OWNER_EXECUTE
+      GROUP_READ GROUP_EXECUTE
+      WORLD_READ WORLD_EXECUTE
+  )
+
+  ExternalProject_Add(zlib_ep
+    ${EP_COMMON_OPTIONS}
+    URL ${ZLIB_SOURCE_URL}
+    URL_HASH "SHA256=${ARROW_ZLIB_BUILD_SHA256_CHECKSUM}"
+    PATCH_COMMAND /bin/bash ${CMAKE_BINARY_DIR}/fix_fdopen_patch.sh
+    BUILD_BYPRODUCTS "${ZLIB_STATIC_LIB}"
+    CMAKE_ARGS ${ZLIB_CMAKE_ARGS}
+  )
+
 
   file(MAKE_DIRECTORY "${ZLIB_PREFIX}/include")
 
